src/inc/til.h

```diff
@@ -18,6 +18,111 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
 #pragma warning(suppress : 26446) // Suppress bounds.4 check for subscript operator.
         return cont[i];
     }
+
+    // color is a universal integral 8bpp RGBA (0-255) color type implicitly convertible to/from
+    // a number of other color types.
+#pragma warning(push)
+    // we can't depend on GSL here (some libraries use BLOCK_GSL), so we use static_cast for explicit narrowing
+#pragma warning(disable : 26472)
+    struct color
+    {
+        uint8_t r, g, b, a;
+
+        constexpr color() noexcept :
+            r{ 0 },
+            g{ 0 },
+            b{ 0 },
+            a{ 0 } {}
+
+        constexpr color(uint8_t _r, uint8_t _g, uint8_t _b) noexcept :
+            r{ _r },
+            g{ _g },
+            b{ _b },
+            a{ 255 } {}
+
+        constexpr color(uint8_t _r, uint8_t _g, uint8_t _b, uint8_t _a) noexcept :
+            r{ _r },
+            g{ _g },
+            b{ _b },
+            a{ _a } {}
+
+        constexpr color(const color&) = default;
+        constexpr color(color&&) = default;
+        color& operator=(const color&) = default;
+        color& operator=(color&&) = default;
+        ~color() = default;
+
+#ifdef _WINDEF_
+        constexpr color(COLORREF c) :
+            r{ static_cast<uint8_t>(c & 0xFF) },
+            g{ static_cast<uint8_t>((c & 0xFF00) >> 8) },
+            b{ static_cast<uint8_t>((c & 0xFF0000) >> 16) },
+            a{ 255 }
+        {
+        }
+
+        operator COLORREF() const noexcept
+        {
+            return static_cast<COLORREF>(r) | (static_cast<COLORREF>(g) << 8) | (static_cast<COLORREF>(b) << 16);
+        }

```

Comment:
What about where we're using `0xffffffff` as `INVALID_COLOR`? How will that work?

Reply:
good point: there is no invalid til::color representation. I don't necessarily think that we _should_ have one; that's more of a `std::optional<til::color>`, right?

Reply:
I suppose, but there are parts of the _console_ codebase particularly that will still need a `0xffffffff` `COLORREF` for `INVALID_COLOR`, that we won't be able to update (I'm thinking esp the propsheet code)